File: modules/room/data/PlayerRoom.ts
import {BaseData} from "../../core/data/BaseData";
import {field} from "../../core/data/DataLoader";
import {PlayerData} from "../../player/data/Player";
import {playerData} from "../../player/managers/PlayerManager";
import {get, Itf} from "../../core/BaseAssist";

export interface IRoomIndex {

	roomId?: string
	npcRoomId?: number

}

export class SkinBuyRecord extends BaseData {

	@field(Number)
	public skinId: number;
	@field(Number)
	public buyTime: number;
	// @field(Number)
	// public price: number;

	public static create(skinId: number) {
		const res = new SkinBuyRecord();
		res.skinId = skinId;
		res.buyTime = Date.now();

		return res;
	}
}

export class RoomCollectRecord extends BaseData {

	@field(String)
	public roomId?: string;
	@field(Number)
	public npcRoomId?: number;
	@field
	public collected: boolean = true;
	@field(Number)
	public changeTime: number;

	public static create(room: IRoomIndex) {
		const res = new RoomCollectRecord();
		res.roomId = room.roomId;
		res.npcRoomId = room.npcRoomId;
		res.changeTime = Date.now();

		return res;
	}
}

export class RoomVisitRecord extends BaseData {

	@field(String)
	public roomId?: string;
	@field(Number)
	public npcRoomId?: number;
	@field(Number)
	public enterTime: number;
	@field(Number)
	public leaveTime?: number;

	public static create(room: IRoomIndex) {
		const res = new RoomVisitRecord();
		res.roomId = room.roomId;
		res.npcRoomId = room.npcRoomId;
		res.enterTime = Date.now();

		return res;
	}

	public leave() {
		this.leaveTime = Date.now();
	}
}

@playerData("PlayerRoom")
export class PlayerRoom extends PlayerData {

	@field([SkinBuyRecord])
	public buyRecords: SkinBuyRecord[] = []
	@field([RoomCollectRecord])
	public collectRecords: RoomCollectRecord[] = []
	@field([RoomVisitRecord])
	public visitRecords: RoomVisitRecord[] = []

	protected syncItf = get("room/player_room/get");

	/**
	 * 当前进入的房间
	 */
	public get visitingRoom(): IRoomIndex {
		return this.visitRecords.find(vr => !vr.leaveTime);
	}

	/**
	 * 已购买的皮肤ID列表
	 */
	public get boughtSkinIds() {
		return this.buyRecords.map(br => br.skinId);
	}

	/**
	 * 购买
	 * @param skinId
	 */
	public buy(skinId) {
		const br = this.getBuy(skinId);
		if (br) return // 已经购买
		this.buyRecords.push(SkinBuyRecord.create(skinId))
	}
	public getBuy(skinId) {
		return this.buyRecords.find(br => br.skinId == skinId);
	}

	/**
	 * 设置收藏
	 */
	public collect(room: IRoomIndex | string | number, collected = true) {
		this.getCollect(room, true).collected = collected;
	}
	public getCollect(room: IRoomIndex | string | number, create = false) {
		if (typeof room == "string") room = {roomId: room};
		if (typeof room == "number") room = {npcRoomId: room};

		let res = this.collectRecords.find(cr => // @ts-ignore
			cr.roomId == room.roomId && cr.npcRoomId == room.npcRoomId);

		if (!res && create) {
			res = RoomCollectRecord.create(room)
			this.collectRecords.push(res);
		}
		return res;
	}

	/**
	 * 访问
	 * @param room
	 */
	public enter(room: IRoomIndex | string | number) {
		this.getVisit(room, true);
	}

	/**
	 * 离开
	 * @param room
	 */
	public leave(room: IRoomIndex | string | number) {
		this.getVisit(room)?.leave();
	}

	public getVisit(room: IRoomIndex | string | number, create = false) {
		if (typeof room == "string") room = {roomId: room};
		if (typeof room == "number") room = {npcRoomId: room};

		let res = this.visitRecords.find(cr => // @ts-ignore
			cr.roomId == room.roomId && cr.npcRoomId == room.npcRoomId);

		if (!res && create) {
			res = RoomVisitRecord.create(room)
			this.visitRecords.push(res);
		}
		return res;
	}


}

File: Learning/project/Go/homi-server/model/playerdata/playerroom.go
package playerdata

import (
	"github.com/go-errors/errors"
	"homi-server-single/common/errs"
	"homi-server-single/models/room"
	"log"
	"time"
)

type PlayerRoom struct {
	PlayerData     `json:",inline" bson:",inline"`
	BuyRecords     []*SkinBuyRecord     `json:"buyRecords" bson:"buyRecords"`
	CollectRecords []*RoomCollectRecord `json:"collectRecords" bson:"collectRecords"`
	VisitRecords   []*RoomVisitRecord   `json:"visitRecords" bson:"visitRecords"`
}

// region 基本数据操作

//func (p PlayerRoom) GetID() string {
//	return p.ID
//}

//func (p PlayerRoom) GetCollection() *mongo.Collection {
//	return oprs.MongoOpr.Collection("PlayerRoom")
//}

//func (p PlayerRoom) Save() error {
//	err := myMongo.Save(p)
//	if err != nil {
//		return err
//	}
//	return nil
//}

//endregion

//region 用户数据处理

//func PlayerRoomData(openid string) (*PlayerRoom, error) {
//	room, err := PlayRoomRepo.FindOne(bson.D{{"openid", openid}})
//	if err != nil {
//		return nil, err
//	}
//	return pr, nil
//}

func (p *PlayerRoom) HasBought(skinId int64) bool {
	result := p.ExtractBoughtSkinRecord(skinId)
	if result == nil { //如果没有找到
		return false //没有购买
	}
	return true
}

func (p *PlayerRoom) GainSkin(skinId int64) error {
	hasBought := p.HasBought(skinId)
	if hasBought {
		log.Println("[Gain Skin] Error", "重复购买房间皮肤")
		return errors.New(errs.HasBoughtRoomErr)
	}
	err := p.CreateSkinBuyRecord(skinId)
	if err != nil {
		return err
	}
	return nil
}

//endregion

func NewPlayerRoom(openid string) *PlayerRoom {
	return &PlayerRoom{PlayerData{Openid: openid}, nil, nil, nil}
}

type RoomCollectRecord struct {
	RoomId     string `json:"roomId"`
	NpcRoomId  int64  `json:"npcRoomId"`
	Collected  bool   `json:"collected"`
	ChangeTime int64  `json:"changeTime"`
}

type RoomVisitRecord struct {
	RoomId    string `json:"roomId"`
	NpcRoomId int64  `json:"npcRoomId"`
	EnterTime int64  `json:"enterTime"`
	LeaveTime int64  `json:"leaveTime"`
}

type SkinBuyRecord struct {
	SkinId  int64 `json:"skinId"`
	BuyTime int64 `json:"buyTime" gorm:"autoCreateTime:milli"`
}

func (p *PlayerRoom) CreateSkinBuyRecord(skinId int64) {
	if p.BuyRecords == nil {
		p.BuyRecords = make([]*SkinBuyRecord, 10)
	}
	p.BuyRecords = append(p.BuyRecords, &SkinBuyRecord{
		SkinId:  skinId,
		BuyTime: time.Now().UnixMilli(),
	})
}

func (p *PlayerRoom) ExtractBoughtSkinRecord(skinId int64) *SkinBuyRecord {
	if p.BuyRecords == nil {
		return nil
	}
	for _, v := range p.BuyRecords {
		if v.SkinId == skinId {
			return v
		}
	}
	return nil
}

func (p *PlayerRoom) SwitchSkin(skinId int64) error {
	roomByOpenid, err := room.GetByOpenId(p.Openid)
	if err != nil {
		return err
	}
	err = roomByOpenid.SwitchSkin(skinId) //获取到对应的Room进行切换
	if err != nil {
		return err
	}
	return nil
}

func (p *PlayerRoom) CreateRoomCollectRecord(idx *room.IRoomIndex, collect bool) {
	record := &RoomCollectRecord{
		RoomId:     idx.RoomId,
		NpcRoomId:  idx.NpcRoom,
		Collected:  collect,
		ChangeTime: time.Now().UnixMilli(),
	}
	if p.CollectRecords == nil {
		p.CollectRecords = make([]*RoomCollectRecord, 10)
	}
	p.CollectRecords = append(p.CollectRecords, record)
}

//// GetVisiting 获取正在访问的房间访问记录
//func GetVisiting(openid string) *RoomVisitRecord {
//	target := &RoomVisitRecord{}
//
//	return target
//}
